src/Ledger/Conway/Script/Validation.agda

```diff
@@ -0,0 +1,131 @@
+
+data ScriptPurpose : Type where
+  Cert     : DCert        → ScriptPurpose
+  Rwrd     : RwdAddr      → ScriptPurpose
+  Mint     : ScriptHash   → ScriptPurpose
+  Spend    : TxIn         → ScriptPurpose
+  Vote     : Voter        → ScriptPurpose
+  Propose  : GovProposal  → ScriptPurpose
+
+rdptr : TxBody → ScriptPurpose → Maybe RdmrPtr
+rdptr txb = λ where
+  (Cert h)     → map (Cert    ,_) $ indexOfDCert    h txcerts
+  (Rwrd h)     → map (Rewrd   ,_) $ indexOfRwdAddr  h txwdrls
+  (Mint h)     → map (Mint    ,_) $ indexOfPolicyId h (policies mint)
+  (Spend h)    → map (Spend   ,_) $ indexOfTxIn     h txins
+  (Vote h)     → map (Vote    ,_) $ indexOfVote     h (map GovVote.voter txvote)
+  (Propose h)  → map (Propose ,_) $ indexOfProposal h txprop
+ where open TxBody txb
+
+indexedRdmrs : Tx → ScriptPurpose → Maybe (Redeemer × ExUnits)
+indexedRdmrs tx sp = maybe (λ x → lookupᵐ? txrdmrs x) nothing (rdptr body sp)
+  where open Tx tx; open TxWitnesses wits
+
+getDatum : Tx → UTxO → ScriptPurpose → Maybe Datum
+getDatum tx utxo (Spend txin) =
+  do (_ , _ , just d , _) ← lookupᵐ? utxo txin
+                          where
+                            (_ , _ , nothing , _) → nothing
+     case d of λ where
+       (inj₁ d) → just d
+       (inj₂ h) → lookupᵐ? m h
+     where
+       m = setToMap (mapˢ < hash , id > (TxWitnesses.txdats (Tx.wits tx)))
+getDatum tx utxo _ = nothing
+
+record TxInfo : Type where
+  field realizedInputs : UTxO
+        txouts  : Ix ⇀ TxOut
+        fee     : Value
+        mint    : Value
+        txcerts : List DCert
+        txwdrls : Wdrl
+        txvldt  : Maybe Slot × Maybe Slot
+        vkKey   : ℙ KeyHash
+        txdats  : ℙ Datum
+        txid    : TxId
+
+txInfo : Language → PParams
+                  → UTxO
+                  → Tx
+                  → TxInfo
+txInfo l pp utxo tx = record
+  { TxBody body
+  ; TxWitnesses wits
+  ; realizedInputs = utxo ∣ txins
+  ; fee = inject txfee
+  ; mint = mint
+  ; vkKey = reqSigHash
+  } where open Tx tx; open TxBody body
+
+credsNeeded : UTxO → TxBody → ℙ (ScriptPurpose × Credential)
+credsNeeded utxo txb
+  =  mapˢ (λ (i , o)  → (Spend  i , payCred (proj₁ o))) ((utxo ∣ (txins ∪ collateral)) ˢ)
+  ∪  mapˢ (λ a        → (Rwrd   a , stake a)) (dom ∣ txwdrls ∣)
+  ∪  mapPartial (λ c  → (Cert   c ,_) <$> cwitness c) (fromList txcerts)
+  ∪  mapˢ (λ x        → (Mint   x , ScriptObj x)) (policies mint)
+  ∪  mapˢ (λ v        → (Vote   v , proj₂ v)) (fromList (map voter txvote))
+  ∪  mapPartial (λ p → if p .policy then (λ {sh} → just (Propose  p , ScriptObj sh)) else nothing)
+                (fromList txprop)
+  where
+    open TxBody txb
+    open GovVote
+    open RwdAddr
+    open GovProposal
+
+scriptsNeeded : UTxO → TxBody → ℙ (ScriptPurpose × ScriptHash)
+scriptsNeeded = mapPartial (λ (sp , c) → if isScriptObj c then (λ {sh} → just (sp , sh)) else nothing)
+                ∘₂ credsNeeded
+
+vKeysNeeded : UTxO → TxBody → ℙ (ScriptPurpose × KeyHash)
```

Comment:
Did you give this a `ScriptPurpose` for symmetry purposes? I can't imagine it ever being used for anything.